File: 98. Validate Binary Search Tree/sol.py
from Typing import Optional 

class TreeNode:
    def __inti__(self, val: int = 0, left : 'Optional[TreeNode]' = None, right: 'Optional[TreeNode]' = None):
        self.val = val 
        self.left = left 
        self.right = right 


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:

        def dfs(node: Optional[TreeNode], highVal: Optional[int] = None, lowVal: Optional[int] = None):
            if not node:
                return True
            if lowVal != None and node.val <= lowVal:
                return False
            if highVal != None and highVal <= node.val:
                return False
            return dfs(node.left, highVal = node.val, lowVal = lowVal) and dfs(node.right, highVal = highVal, lowVal = node.val)

        return dfs(root)

File: 98. Validate Binary Search Tree/sol.go
package validateBinarySearchTree

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isValidBST(root *TreeNode) bool {
	var dfs func(node *TreeNode, highVal, lowVal *int) bool
	dfs = func(node *TreeNode, highVal, lowVal *int) bool {
		if node == nil {
			return true
		}
		if lowVal != nil && *lowVal >= node.Val {
			return false
		}
		if highVal != nil && node.Val >= *highVal {
			return false
		}
		return dfs(node.Left, &node.Val, lowVal) && dfs(node.Right, highVal, &node.Val)
	}
	return dfs(root, nil, nil)
}

// BFS
func isValidBSTBFS(root *TreeNode) bool {
	que := deque{}
	que.append(root, nil, nil)
	for len(que.nodes) > 0 {
		node, low, high := que.popleft()
		if low != nil && *low >= node.Val {
			return false
		}
		if high != nil && *high <= node.Val {
			return false
		}
		if node.Left != nil {
			que.append(node.Left, low, &node.Val)
		}
		if node.Right != nil {
			que.append(node.Right, &node.Val, high)
		}

	}

	return true

}

type deque struct {
	nodes []que
}

type que struct {
	node *TreeNode
	low  *int
	high *int
}

func (q *deque) init() *deque {
	q.nodes = make([]que, 0)
	return q
}

func (q *deque) append(node *TreeNode, low, high *int) {

	q.nodes = append(q.nodes, que{node, low, high})
}

func (q *deque) popleft() (*TreeNode, *int, *int) {
	if len(q.nodes) > 0 {
		val := q.nodes[0]
		q.nodes = q.nodes[1:]
		return val.node, val.low, val.high
	}
	return nil, nil, nil
}
